drivers/wifi/infineon/cyw43xxx_drv.c

```diff
@@ -0,0 +1,713 @@
+		memcpy(&whd_scan_result, *result_ptr, sizeof(whd_scan_result_t));
+		parse_scan_result(&whd_scan_result, &zephyr_scan_result);
+		data->scan_rslt_cb(data->iface, 0, &zephyr_scan_result);
+	}
+	memset(*result_ptr, 0, sizeof(whd_scan_result_t));
+}
+
+static int cyw43xxx_mgmt_scan(const struct device *dev, struct wifi_scan_params *params,
+			      scan_result_cb_t cb)
+{
+	int ret = 0;
+	struct cyw43xxx_wifi_runtime *data = (struct cyw43xxx_wifi_runtime *)dev->data;
+
+	data->scan_rslt_cb = cb;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(CYW43XXX_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	ret = whd_wifi_scan(cyw43xxx_if, params->scan_type, WHD_BSS_TYPE_ANY, &(data->ssid), NULL,
+			    NULL, NULL, scan_callback, &(data->scan_result), data);
+
+	k_sem_give(&data->sema_common);
+	return ret;
+}
+
+static int cyw43xxx_mgmt_connect(const struct device *dev, struct wifi_connect_req_params *params)
+{
+	struct cyw43xxx_wifi_runtime *data = (struct cyw43xxx_wifi_runtime *)dev->data;
+	whd_ssid_t ssid = {0};
+	int ret = 0;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(CYW43XXX_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	if (data->is_sta_connected) {
+		LOG_ERR("Already connected");
+		ret = -EALREADY;
+		goto error;
+	}
+
+	ssid.length = params->ssid_length;
+	memcpy(ssid.value, params->ssid, params->ssid_length);
+
+	whd_scan_result_t scan_result;
+	whd_scan_result_t usr_result = {0};
+
+	usr_result.SSID.length = ssid.length;
+	memcpy(usr_result.SSID.value, ssid.value, ssid.length);
+
+	if (whd_wifi_scan(cyw43xxx_if, WHD_SCAN_TYPE_ACTIVE, WHD_BSS_TYPE_ANY, NULL, NULL, NULL,
+			  NULL, scan_cb_search, &scan_result, &(usr_result)) != WHD_SUCCESS) {
+		LOG_ERR("Failed start scan");
+		ret = -EAGAIN;
+		goto error;
+	}
+
+	if (k_sem_take(&cyw43xxx_wifi_data.sema_scan, K_MSEC(12 * 1000)) != 0) {
+		whd_wifi_stop_scan(cyw43xxx_if);
+		ret = -EAGAIN;
+		goto error;
+	}
+
+	if (usr_result.security == 0) {
+		ret = -EAGAIN;
+		goto error;
+	}
+
+	/* Connect to the network */
+	if (whd_wifi_join(cyw43xxx_if, &usr_result.SSID, usr_result.security, params->psk,
+			  params->psk_length) != WHD_SUCCESS) {
+		LOG_ERR("Failed to connect with network");
+		ret = -EAGAIN;
+		goto error;
+	}
+
+error:
+	if (ret < 0) {
+		net_if_dormant_on(data->iface);
+	} else {
+		net_dhcpv4_start(data->iface);
+		net_if_dormant_off(data->iface);
+		data->is_sta_connected = 1;
+	}
+
+	wifi_mgmt_raise_connect_result_event(data->iface, ret);
+	k_sem_give(&data->sema_common);
+	return ret;
+}
+
+static int cyw43xxx_mgmt_disconnect(const struct device *dev)
+{
+	int ret = 0;
+	struct cyw43xxx_wifi_runtime *data = (struct cyw43xxx_wifi_runtime *)dev->data;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(CYW43XXX_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	if (whd_wifi_leave(cyw43xxx_if) != WHD_SUCCESS) {
+		k_sem_give(&data->sema_common);
+		ret = -EAGAIN;
+	} else {
+		data->is_sta_connected = 0;
+		net_dhcpv4_stop(data->iface);
+		net_if_dormant_on(data->iface);
```

Comment:
Is there an asynchronous "disconnected" event somewhere that should also be setting dormant back to on? I can only see that happening here and if `cyw43xxx_mgmt_connect` fails.

Reply:
added net_if_dormant_on to cyw43xxx_event_task